### PR TITLE
Update plugin for General Data Protection Regulation

### PR DESCRIPTION
**Pull request by NinaHerrmann**
This Pull request is part of fixing Issue [5] (https://github.com/learnweb/his_unification/issues/5). 
However, deletion of the data is still missing.

### CODE REVIEW

**Comment by Dagefoerde** [comment]
local_lsf_unification's function do actually get called, but I get this error when trying to export

```
    Processing local_lsf_unification (423/491) (Saturday, 26 May 2018, 11:39 AM)
++ ERROR: Incorrect number of query parameters. Expected 2, got 1. ++
* line 741 of /lib/dml/pgsql_native_moodle_database.php: call to moodle_database->fix_sql_params()
* line 77 of /local/lsf_unification/classes/privacy/provider.php: call to pgsql_native_moodle_database->get_recordset_sql()
* line 76 of /privacy/classes/local/legacy_polyfill.php: call to local_lsf_unification\privacy\provider::_get_contexts_for_userid()
* line 7757 of /lib/moodlelib.php: call to local_lsf_unification\privacy\provider::get_contexts_for_userid()
* line 495 of /privacy/classes/manager.php: call to component_class_callback()
* line 528 of /privacy/classes/manager.php: call to core_privacy\manager::component_class_callback()
* line 237 of /privacy/classes/manager.php: call to core_privacy\manager->handled_component_class_callback()
* line 70 of /export.php: call to core_privacy\manager->get_contexts_for_userid()
```

The problem is https://github.com/learnweb/his_unification/pull/6/files#diff-66123c1486cf3b72c27270d008e174b2R74. There are two parameters (with the same name, but the DB layer doesn't seem to care).  For that reason the contexts are not assembled correctly.

**Comment by Dagefoerde** [comment]
once I fix the params, everything (i.e. the dummy output) works right away, see this screenshot:

output

File: lang/de/local_lsf_unification.php [hunk 1]
```diff
@@ -166,4 +166,13 @@
...
-
+// Privacy API
+$string['privacy:metadata:local_lsf_unification:veranstid'] = 'ID der Veranstaltung im HisLSF System';
+$string['privacy:metadata:local_lsf_unification:mdlid'] = 'ID der Veranstaltung im Moodle System';
+$string['privacy:metadata:local_lsf_unification:timestamp'] = 'Zeitpunkt yu dem der Kurs importiert wurde';
+$string['privacy:metadata:local_lsf_unification:requeststate'] = 'Status der Kursanfrage';
+$string['privacy:metadata:local_lsf_unification:requesterid'] = 'ID der Person die den Kurs angefragt hat';
+$string['privacy:metadata:local_lsf_unification:acceptorid'] = 'ID der Person die den Kurs akzeptiert hat';
+
+$string['privacy:local_lsf_unification:requester'] = 'Die Anfrage f&uuml;r einen Kurs wurde gespeichert.';
```
Comment (Dagefoerde) [inline lang/de/local_lsf_unification.php]: This string, and the following, and the corresponding ones in the english version are not used.

File: lang/de/local_lsf_unification.php [hunk 2]
```diff
@@ -166,4 +166,13 @@
 $string['eventcourse_imported'] = 'Kurs importiert';
 $string['eventmatchingtable_updated'] = 'Matchingtable ge&auml;ndert';
 $string['eventcourse_duplicated'] = 'Kursinhalte dupliziert';
-
+// Privacy API
+$string['privacy:metadata:local_lsf_unification:veranstid'] = 'ID der Veranstaltung im HisLSF System';
+$string['privacy:metadata:local_lsf_unification:mdlid'] = 'ID der Veranstaltung im Moodle System';
+$string['privacy:metadata:local_lsf_unification:timestamp'] = 'Zeitpunkt yu dem der Kurs importiert wurde';
```
Comment (Dagefoerde) [inline lang/de/local_lsf_unification.php]: "yu" ;)

File: lang/de/local_lsf_unification.php [hunk 3]
```diff
@@ -166,4 +166,13 @@
 $string['eventcourse_imported'] = 'Kurs importiert';
 $string['eventmatchingtable_updated'] = 'Matchingtable ge&auml;ndert';
 $string['eventcourse_duplicated'] = 'Kursinhalte dupliziert';
-
+// Privacy API
+$string['privacy:metadata:local_lsf_unification:veranstid'] = 'ID der Veranstaltung im HisLSF System';
```
Comment (Dagefoerde) [inline lang/de/local_lsf_unification.php]: Sollte sein: HISLSF (auch im englischen)

File: classes/privacy/provider.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                    "is waiting";
+                break;
+            case 2 :
+                $status =
+                    "is granted";
+                break;
+        }
+        $data = (object)[
+            'action' => get_string('privacy:local_lsf_unification:' . $action, 'local_lsf_unification'),
+            'timestamp' => date('m/d/Y H:i:s', $courserequest->timestamp),
```
Comment (Dagefoerde) [inline classes/privacy/provider.php]: Ich bin gegen amerikanisches Datumsformat, das kann doch kein vernünftiger Mensch lesen ;)  Moodle hat Funktionen, um Datumsangaben genau so auszugeben, wie ein Nutzer sie lesen kann. Benutz lieber die, oder direkt den `$timestamp`.

File: lang/de/local_lsf_unification.php (hunk 1 above)
Comment (Dagefoerde) [reply lang/de/local_lsf_unification.php]: Ohhhh, they are, but indirectly. Sorry. Hm, that could be clarified in a comment.  Acceptor could be explained as 
"Eine Kursanfrage wurde akzeptiert" instead.

File: classes/privacy/provider.php [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+     */
+    public static function _export_user_data(approved_contextlist $contextlist) {
+        global $DB;
+
+        if (empty($contextlist->count())) {
+            return;
+        }
+
+        $user = $contextlist->get_user();
+        $context = \context_user::instance($contextlist->get_user()->id);
```
Comment (Dagefoerde) [inline classes/privacy/provider.php]: `$user->id`

File: classes/privacy/provider.php (hunk 4 above)
Comment (Dagefoerde) [reply classes/privacy/provider.php]: .. oder `date('c', $timestamp) für ISO 8601-konformes Datum.

File: classes/privacy/provider.php [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Deletes data from the table for all users in all contexts.
+     * @param $context
+     * @throws \dml_exception
+     */
+    public static function _delete_data_for_all_users_in_context($context) {
+        global $DB;
+
+        // Sanity check that context is at the System context level.
```
Comment (Dagefoerde) [inline classes/privacy/provider.php]: Comment is not accurate anymore.

In addition, also needs code for the case that `$context` is the system context; i.e. purge everything.

File: lang/de/local_lsf_unification.php (hunk 1 above)
Comment (NinaHerrmann) [reply lang/de/local_lsf_unification.php]: The course can also be declined therefore I chose for the cumbersome formulation.

**Pull request merged**